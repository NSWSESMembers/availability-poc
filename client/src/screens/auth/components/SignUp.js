import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Container, Holder } from '../../../components/Container';
import { Input } from '../../../components/Input';
import { Button } from '../../../components/Button';
import { Progress } from '../../../components/Progress';

class SignUp extends Component {
  state = {
    username: '',
    password: '',
    email: '',
  };

  onChangeUsername = (value) => {
    this.setState({
      username: value,
    });
  };

  onChangeEmail = (value) => {
    this.setState({
      email: value,
    });
  };

  onChangePassword = (value) => {
    this.setState({
      password: value,
    });
  };

  onSignUp = () => {
    const { username, email, password } = this.state;
    this.props.onSignUp({ username, email, password });
  }

  render() {
    const { loading } = this.props;
    const { username, email, password } = this.state;
    return (
      <Container>
        <Holder margin transparent>
          <Input
            type="username"
            placeholder="Username"
            value={username}
            onChangeText={this.onChangeUsername}
          />
          <Input
            type="password"
            placeholder="Password"
            value={password}
            onChangeText={this.onChangePassword}
          />
          <Input
            type="email"
            placeholder="Email"
            value={email}
            onChangeText={this.onChangeEmail}
          />
          <Button text="Register" onPress={this.onSignUp} disabled={loading} />
        </Holder>
        {loading && (
          <Container>
            <Progress />
          </Container>
        )}
      </Container>
    );
  }
}

SignUp.propTypes = {
  onSignUp: PropTypes.func.isRequired,
  loading: PropTypes.bool,
};

export default SignUp;
