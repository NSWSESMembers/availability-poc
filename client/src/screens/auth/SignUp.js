import React, { Component } from 'react';
import { Keyboard } from 'react-native';
import { connect } from 'react-redux';
import { graphql, compose } from 'react-apollo';
import PropTypes from 'prop-types';

import { setCurrentUser } from '../../state/auth.actions';

import { Container, Holder } from '../../components/Container';
import { Header } from '../../components/Header';
import { Input } from '../../components/TextInput';
import { NavBar } from '../../components/NavBar';
import { NextButton } from '../../components/Button';
import { Alert } from '../../components/Alert';

import SIGNUP_MUTATION from '../../graphql/signup.mutation';

class SignUp extends Component {
  state = {
    username: '',
    password: '',
    email: '',
    usernameValid: false,
    passwordValid: false,
    emailValid: false,
  }

  onPressSignUp = () => {
    Keyboard.dismiss();
    const { deviceUuid } = this.props.local;
    const { username, password, email } = this.state;
    this.props
      .signup({ username, email, password, deviceUuid })
      .then(({ data: { signup: user } }) => {
        const ourUser = {
          username: user.username,
          token: user.authToken,
        };
        this.props.dispatch(setCurrentUser(ourUser));
      })
      .catch((error) => {
        this.setState({
          status: 'Sign Up Error',
          errorMessage: error.message,
        });
        this.popRef.show();
      });
  }

  onChangeEmail = (value) => {
    let emailValid = false;
    if (value.indexOf('@') !== -1) emailValid = true;

    this.setState({
      email: value,
      emailValid,
    });
  };

  onChangePassword = (value) => {
    let passwordValid = false;
    if (value.length > 5) passwordValid = true;

    this.setState({
      password: value,
      passwordValid,
    });
  };

  onChangeUsername = (value) => {
    let usernameValid = false;
    if (value.length > 5) usernameValid = true;

    this.setState({
      username: value,
      usernameValid,
    });
  };

  onPressBack = () => {
    this.props.navigation.goBack();
  };

  render() {
    return (
      <Container>
        <NavBar onPressBack={this.onPressBack} />
        <Holder>
          <Header title="Sign Up" />
          <Input
            title="Username"
            onChangeText={this.onChangeUsername}
            valid={this.state.usernameValid}
            type="username"
          />
          <Input
            title="Password"
            onChangeText={this.onChangePassword}
            valid={this.state.passwordValid}
            type="password"
          />
          <Input
            title="Email"
            onChangeText={this.onChangeEmail}
            valid={this.state.emailValid}
            type="email"
          />
        </Holder>
        <NextButton
          onPress={this.onPressSignUp}
          disabled={
            !(this.state.usernameValid && this.state.passwordValid && this.state.emailValid)
          }
        />
        <Alert
          ref={(el) => {
            this.popRef = el;
          }}
          status={this.state.status}
          message={this.state.errorMessage}
        />
      </Container>
    );
  }
}

const mapStateToProps = ({ auth, local }) => ({
  auth,
  local,
});

SignUp.propTypes = {
  navigation: PropTypes.shape({
    goBack: PropTypes.func,
  }),
  dispatch: PropTypes.func.isRequired,
  local: PropTypes.shape({
    deviceUuid: PropTypes.string,
  }),
  signup: PropTypes.func.isRequired,
};

const signup = graphql(SIGNUP_MUTATION, {
  props: ({ mutate }) => ({
    signup: ({ username, email, password, deviceUuid }) =>
      mutate({
        variables: { user: { username, email, password, deviceUuid } },
      }),
  }),
});

export default compose(signup, connect(mapStateToProps))(SignUp);
