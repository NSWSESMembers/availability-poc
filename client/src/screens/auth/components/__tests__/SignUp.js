import React from 'react';
import TestRenderer from 'react-test-renderer';

import SignUp from '../SignUp';


test('renders correctly', () => {
  const onSignUp = jest.fn();
  const tree1 = TestRenderer.create(
    <SignUp
      onSignUp={onSignUp}
    />,
  ).toJSON();
  expect(tree1).toMatchSnapshot();

  const tree2 = TestRenderer.create(
    <SignUp
      onSignUp={onSignUp}
      loading
    />,
  ).toJSON();
  expect(tree2).toMatchSnapshot();
});

test('events', () => {
  const username = 'username';
  const password = 'password';
  const email = 'email';
  const onSignUp = jest.fn();
  const renderer = TestRenderer.create(
    <SignUp
      onSignUp={onSignUp}
      loading={false}
    />,
  );

  const signUp = renderer.getInstance();
  signUp.onChangeUsername(username);
  signUp.onChangePassword(password);
  signUp.onChangeEmail(email);
  signUp.onSignUp();

  expect(onSignUp.mock.calls.length).toBe(1);
  expect(onSignUp.mock.calls[0][0]).toEqual({ username, email, password });
});
