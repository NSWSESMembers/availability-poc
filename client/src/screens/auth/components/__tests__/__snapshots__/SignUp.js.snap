// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#EEE",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "paddingLeft": 20,
          "paddingRight": 20,
          "width": "100%",
        },
        Object {
          "paddingTop": 20,
        },
        Object {
          "backgroundColor": "transparent",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "borderBottomWidth": 0.5,
          "borderColor": "#cbcbcb",
          "marginBottom": 25,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          onChangeText={[Function]}
          placeholder="Username"
          placeholderTextColor="#C2C2C2"
          returnKeyType="next"
          style={
            Object {
              "alignItems": "center",
              "color": "#1e3c78",
              "flex": 1,
              "flexDirection": "row",
              "fontSize": 20,
              "fontWeight": "700",
              "height": 50,
              "justifyContent": "center",
            }
          }
          underlineColorAndroid="rgba(0,0,0,0)"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "borderBottomWidth": 0.5,
          "borderColor": "#cbcbcb",
          "marginBottom": 25,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          placeholder="Password"
          placeholderTextColor="#C2C2C2"
          secureTextEntry={true}
          style={
            Object {
              "alignItems": "center",
              "color": "#1e3c78",
              "flex": 1,
              "flexDirection": "row",
              "fontSize": 20,
              "fontWeight": "700",
              "height": 50,
              "justifyContent": "center",
            }
          }
          underlineColorAndroid="rgba(0,0,0,0)"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "borderBottomWidth": 0.5,
          "borderColor": "#cbcbcb",
          "marginBottom": 25,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          keyboardType="email-address"
          onChangeText={[Function]}
          placeholder="Email"
          placeholderTextColor="#C2C2C2"
          returnKeyType="next"
          style={
            Object {
              "alignItems": "center",
              "color": "#1e3c78",
              "flex": 1,
              "flexDirection": "row",
              "fontSize": 20,
              "fontWeight": "700",
              "height": 50,
              "justifyContent": "center",
            }
          }
          underlineColorAndroid="rgba(0,0,0,0)"
          value=""
        />
      </View>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#f48603",
          "opacity": 1,
          "padding": 10,
          "width": "100%",
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#ffffff",
              "fontSize": 20,
              "textAlign": "center",
            },
          ]
        }
      >
        Register
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders correctly 2`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#EEE",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "flex-start",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "paddingLeft": 20,
          "paddingRight": 20,
          "width": "100%",
        },
        Object {
          "paddingTop": 20,
        },
        Object {
          "backgroundColor": "transparent",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "borderBottomWidth": 0.5,
          "borderColor": "#cbcbcb",
          "marginBottom": 25,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          onChangeText={[Function]}
          placeholder="Username"
          placeholderTextColor="#C2C2C2"
          returnKeyType="next"
          style={
            Object {
              "alignItems": "center",
              "color": "#1e3c78",
              "flex": 1,
              "flexDirection": "row",
              "fontSize": 20,
              "fontWeight": "700",
              "height": 50,
              "justifyContent": "center",
            }
          }
          underlineColorAndroid="rgba(0,0,0,0)"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "borderBottomWidth": 0.5,
          "borderColor": "#cbcbcb",
          "marginBottom": 25,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          placeholder="Password"
          placeholderTextColor="#C2C2C2"
          secureTextEntry={true}
          style={
            Object {
              "alignItems": "center",
              "color": "#1e3c78",
              "flex": 1,
              "flexDirection": "row",
              "fontSize": 20,
              "fontWeight": "700",
              "height": 50,
              "justifyContent": "center",
            }
          }
          underlineColorAndroid="rgba(0,0,0,0)"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "borderBottomWidth": 0.5,
          "borderColor": "#cbcbcb",
          "marginBottom": 25,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          keyboardType="email-address"
          onChangeText={[Function]}
          placeholder="Email"
          placeholderTextColor="#C2C2C2"
          returnKeyType="next"
          style={
            Object {
              "alignItems": "center",
              "color": "#1e3c78",
              "flex": 1,
              "flexDirection": "row",
              "fontSize": 20,
              "fontWeight": "700",
              "height": 50,
              "justifyContent": "center",
            }
          }
          underlineColorAndroid="rgba(0,0,0,0)"
          value=""
        />
      </View>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#f48603",
          "opacity": 1,
          "padding": 10,
          "width": "100%",
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#ffffff",
              "fontSize": 20,
              "textAlign": "center",
            },
          ]
        }
      >
        Register
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#EEE",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </View>
</View>
`;
